File: Android/app/src/main/java/leapfrog_inc/icchi/Http/Requester/AccountRequester.java
package leapfrog_inc.icchi.Http.Requester;

import android.util.Base64;

import org.json.JSONObject;

import leapfrog_inc.icchi.Function.Constants;
import leapfrog_inc.icchi.Http.HttpManager;

/**
 * Created by Leapfrog-Software on 2018/01/27.
 */

public class AccountRequester {

    public static void register(String email, String password, String name, String image, String fbLink, final RegisterCallback callback){

        HttpManager http = new HttpManager(new HttpManager.HttpCallback() {
            @Override
            public void didReceiveData(boolean result, String data) {
                if (result) {
                    try {
                        JSONObject jsonObject = new JSONObject(data);
                        String ret = jsonObject.getString("result");
                        String userId = jsonObject.getString("userId");
                        callback.didReceive(ret.equals("0"), userId);
                        return;
                    } catch(Exception e) {}
                }
                callback.didReceive(false, null);
            }
        });
        StringBuffer param = new StringBuffer();
        param.append("command=register");
        param.append("&");
        param.append("email=" + email);
        param.append("&");
        param.append("password=" + password);
        param.append("&");
        param.append("name=" + Base64.encodeToString(name.getBytes(), Base64.URL_SAFE | Base64.NO_WRAP));
        param.append("&");
        param.append("image=" + image);
        param.append("&");
        param.append("fbLink=" + fbLink);
        http.execute(Constants.ServerRootUrl, "POST", param.toString());
    }

    public interface RegisterCallback {
        void didReceive(boolean result, String userId);
    }

    public static void login(String email, String password, final LoginCallback callback){

        HttpManager http = new HttpManager(new HttpManager.HttpCallback() {
            @Override
            public void didReceiveData(boolean result, String data) {
                if (result) {
                    try {
                        JSONObject jsonObject = new JSONObject(data);
                        String ret = jsonObject.getString("result");
                        String userId = jsonObject.getString("userId");
                        callback.didReceive(ret.equals("0"), userId);
                        return;
                    } catch(Exception e) {}
                }
                callback.didReceive(false, null);
            }
        });
        StringBuffer param = new StringBuffer();
        param.append("command=login");
        param.append("&");
        param.append("email=" + email);
        param.append("&");
        param.append("password=" + password);
        http.execute(Constants.ServerRootUrl, "POST", param.toString());
    }

    public interface LoginCallback {
        void didReceive(boolean result, String userId);
    }

    public static void update(UserRequester.UserData userData, final UpdateCallback callback) {

        if (userData == null) {
            callback.didReceive(false);
            return;
        }

        HttpManager http = new HttpManager(new HttpManager.HttpCallback() {
            @Override
            public void didReceiveData(boolean result, String data) {
                if (result) {
                    try {
                        JSONObject jsonObject = new JSONObject(data);
                        String ret = jsonObject.getString("result");
                        if (ret.equals("0")) {
                            callback.didReceive(true);
                            return;
                        }
                    } catch (Exception e) {}
                }
                callback.didReceive(false);
            }
        });
        StringBuffer param = new StringBuffer();
        param.append("command=updateAccount");
        param.append("&");
        param.append("userId=" + userData.userId);
        param.append("&");

        param.append("name=" + Base64.encodeToString(userData.name.getBytes(), Base64.URL_SAFE | Base64.NO_WRAP));
        param.append("&");
        param.append("age=");
        if (userData.age != null) {
            param.append(userData.age.convert());
        }
        param.append("&");
        param.append("gender=");
        if (userData.gender != null) {
            param.append(userData.gender.convert());
        }
        param.append("&");

        String likes = "";
        for (int i = 0; i < userData.likes.size(); i++) {
            if (likes.length() > 0) {
                likes += "-";
            }
            likes += userData.likes.get(i);
        }
        param.append("likes=" + likes);
        param.append("&");

        String hates = "";
        for (int i = 0; i < userData.hates.size(); i++) {
            if (hates.length() > 0) {
                hates += "-";
            }
            hates += userData.hates.get(i);
        }
        param.append("hates=" + hates);

        param.append("&");
        param.append("image=" + userData.image);
        param.append("&");
        param.append("fbLink=" + userData.fbLink);

        http.execute(Constants.ServerRootUrl, "POST", param.toString());
    }

    public interface UpdateCallback {
        void didReceive(boolean result);
    }
}

File: iOS/Icchi/Icchi/Http/Requester/AccountRequester.swift
//
//  AccountRequester.swift
//  Icchi
//
//  Created by oonaka on 2018/02/22.
//  Copyright © 2018年 Leapfrog-Inc. All rights reserved.
//

import Foundation

class AccountRequester {
    
    // レスポンスデータ型
    private struct RegisterResult : Codable {
        var result:Bool
        var userId:String
        private enum CodingKeys: String, CodingKey {
            case result
            case userId
        }
        init(from decoder: Decoder) throws {
            let values = try decoder.container(keyedBy: CodingKeys.self)
            result = NSString(string:try values.decode(String.self, forKey: .result)).isEqual(to: "0")
            userId = try values.decode(String.self, forKey: .userId)
        }
    }
    // レスポンスデータ型
    private struct UpdateResult : Codable {
        var result:Bool
        private enum CodingKeys: String, CodingKey {
            case result
        }
        init(from decoder: Decoder) throws {
            let values = try decoder.container(keyedBy: CodingKeys.self)
            result = NSString(string:try values.decode(String.self, forKey: .result)).isEqual(to: "0")
        }
    }
    
    // シングルトン
    static let sharedManager = AccountRequester()
    private init() {
    }
    
    /** 登録 */
    public func register(email:String, password:String, name:String, image:String, fbLink:String,  completion:@escaping((Bool, String?)->Void)) -> Void {
        
        // リクエストデータ作成
        let params = ["command": "register"
                    , "email":email
                    , "password":password
                    , "name":name
                    , "image":image
                    , "fbLink":fbLink
                    ]
        // リクエスト実施
        HttpManager.post(url: Constants.ServerApiUrl, params: params) { (result:Bool, data:Data?) in
            
            // レスポンスチェック
            guard result, let data = data else {
                // 失敗
                completion(false, nil)
                return
            }
            
            do {
                // レスポンス取得
                let result = try JSONDecoder().decode(RegisterResult.self, from: data)
                
                // 成功
                completion(result.result, result.userId)
            }
            catch let error {
                // 失敗
                print("AccountRequester regist decode error = \(error)")
                completion(false, nil)
            }
        }
    }
    
    /** ログイン */
    public func login(email:String, password:String, completion:@escaping((Bool, String?)->Void)) -> Void {
        
        // リクエストデータ作成
        let params = ["command": "login"
            , "email":email
            , "password":password
        ]
        // リクエスト実施
        HttpManager.post(url: Constants.ServerApiUrl, params: params) { (result:Bool, data:Data?) in
            
            // レスポンスチェック
            guard result, let data = data else {
                // 失敗
                completion(false, nil)
                return
            }
            
            do {
                // レスポンス取得
                let result = try JSONDecoder().decode(RegisterResult.self, from: data)
                
                // 成功
                completion(result.result, result.userId)
            }
            catch let error {
                // 失敗
                print("AccountRequester login decode error = \(error)")
                completion(false, nil)
            }
        }
    }
    
    func update(userData:UserRequester.UserData?, completion:@escaping((Bool)->Void)) -> Void {
        
        guard let userData = userData else {
            completion(false)
            return
        }
        // リクエストデータ作成
        let params:[String:String] = ["command": "updateAccount"
            , "userId":userData.userId ?? ""
            , "name":userData.name?.base64Encode() ?? ""
            , "age":userData.age?.convert() ?? ""
            , "gender":userData.gender?.convert() ?? ""
            , "likes":userData.likes?.joined(separator: "-") ?? ""
            , "hates":userData.hates?.joined(separator: "-") ?? ""
            , "image":userData.image ?? ""
            , "fbLink":userData.fbLink ?? ""
        ]
        // リクエスト実施
        HttpManager.post(url: Constants.ServerApiUrl, params: params) { (result:Bool, data:Data?) in
            
            // レスポンスチェック
            guard result, let data = data else {
                // 失敗
                completion(false)
                return
            }
            
            do {
                // レスポンス取得
                let result = try JSONDecoder().decode(UpdateResult.self, from: data)
                
                // 成功
                completion(result.result)
            }
            catch let error {
                // 失敗
                print("AccountRequester login decode error = \(error)")
                completion(false)
            }
        }
    }
}
